tools/stubmaker/main.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"go/types"
+	"io"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"golang.org/x/tools/go/packages"
+)
+
+func isEnterprise() bool {
+	repo, err := git.PlainOpenWithOptions(".", &git.PlainOpenOptions{
+		DetectDotGit: true,
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+	wt, err := repo.Worktree()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	st, err := wt.Filesystem.Stat("enthelpers")
+	onOss := errors.Is(err, os.ErrNotExist)
+	onEnt := st != nil
+
+	switch {
+	case onOss && !onEnt:
+	case !onOss && onEnt:
+	default:
+		log.Fatal(err)
+	}
+	return onEnt
+}
+
+func main() {
+	if !isEnterprise() {
+		return
+	}
+	baseFilename := strings.TrimSuffix(os.Getenv("GOFILE"), "_oss.go")
+	if err := writeStubIfNeeded(baseFilename); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func readLines(r io.Reader) ([]string, error) {
+	scanner := bufio.NewScanner(r)
+	scanner.Split(bufio.ScanLines)
+
+	var lines []string
+	for scanner.Scan() {
+		lines = append(lines, scanner.Text())
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return lines, nil
+}
+
+func writeStubIfNeeded(baseFilename string) (err error) {
+	warning := "// Code generated by tools/stubmaker; DO NOT EDIT."
+
+	var output *os.File
+	inputFile := fmt.Sprintf("%s_oss.go", baseFilename)
+	b, err := os.ReadFile(inputFile)
+	if err != nil {
+		return err
+	}
+
+	inputLines, err := readLines(bytes.NewBuffer(b))
+	var funcs []string
+	var outputLines []string
+	for _, line := range inputLines {
+		switch line {
+		case "//go:build !enterprise":
+			outputLines = append(outputLines, warning, "")
+			line = "//go:build enterprise"
+		case "//go:generate go run github.com/hashicorp/vault/tools/stubmaker":
+			continue
+		}
+		outputLines = append(outputLines, line)
+
+		trimmed := strings.TrimSpace(line)
+		if strings.HasPrefix(trimmed, "func ") {
+			i := strings.Index(trimmed, "(")
+			if i != -1 {
+				funcs = append(funcs, trimmed[5:i])
+			}
+		}
+	}
+
+	target := fmt.Sprintf("%s_ent.go", baseFilename)
+	os.Remove(target)
```

Comment:
do we want to remove the ent file, if it exists?

Reply:
I think we do.  It's possible that we created it in a previous run, but since then the real funcs were implemented, and now the presence of the file causes builds to fail due to multiple definitions of the funcs.